projects/charger/src/notify.c

```diff
@@ -0,0 +1,85 @@
+#include "notify.h"
+
+#include <stddef.h>
+
+#include "can_interval.h"
+#include "can_msg_defs.h"
+#include "can_pack.h"
+#include "can_unpack.h"
+#include "charger_events.h"
+#include "event_queue.h"
+#include "exported_enums.h"
+#include "generic_can.h"
+#include "generic_can_msg.h"
+#include "log.h"
+#include "soft_timer.h"
+#include "status.h"
+
+static uint32_t s_watchdog_period_s;
+static CanInterval *s_interval;
+static SoftTimerID s_watchdog = SOFT_TIMER_INVALID_TIMER;
+
+// Watchdog activated when commanded to start. Requires reaffirmation every minute or it will cease
+// charging.
+static void prv_command_watchdog(SoftTimerID id, void *context) {
+  (void)id;
+  (void)context;
+  event_raise(CHARGER_EVENT_STOP_CHARGING, 0);
+}
+
+static void prv_kick_watchdog(void) {
+  soft_timer_cancel(s_watchdog);
+  soft_timer_start_seconds(s_watchdog_period_s, prv_command_watchdog, NULL, &s_watchdog);
+}
+
+// GenericCanRx
+static void prv_command_rx(const GenericCanMsg *msg, void *context) {
+  (void)context;
+  CANMessage can_msg = { 0 };
+  generic_can_msg_to_can_message(msg, &can_msg);
+  EEChargerSetRelayState relay_state = 0;
+  CAN_UNPACK_CHARGER_SET_RELAY_STATE(&can_msg, (uint8_t *)&relay_state);
+  if (relay_state == EE_CHARGER_SET_RELAY_STATE_CLOSE) {
+    event_raise(CHARGER_EVENT_START_CHARGING, 0);
```

Comment:
This doesn't seem to have any protection against attempting to charge while the charger is disconnected. Is that something you plan on handling in an FSM?

Reply:
Yes the FSM takes care of this. `CHARGER_EVENT_START_CHARGING` will only work in the connected state. The charger has 3 states:

- Disconnected
- Connected
- Charging